SYCL/AtomicRef/add_generic.cpp

```diff
@@ -1,43 +1,157 @@
+// RUN: %HOST_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+
+// RUN: %clangxx -fsycl -fsycl-unnamed-lambda -fsycl-targets=%sycl_triple %s -o %t.out -Xsycl-target-backend=nvptx64-nvidia-cuda --cuda-gpu-arch=sm_70 -DRELEASE -DSYSTEM
+// RUN: %HOST_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+
+// RUN: %clangxx -fsycl -fsycl-unnamed-lambda -fsycl-targets=%sycl_triple %s -o %t.out -Xsycl-target-backend=nvptx64-nvidia-cuda --cuda-gpu-arch=sm_70 -DRELEASE -DSYSTEM -DATOMIC64
+// RUN: %HOST_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+
+// RUN: %clangxx -fsycl -fsycl-unnamed-lambda -fsycl-targets=%sycl_triple %s -o %t.out -Xsycl-target-backend=nvptx64-nvidia-cuda --cuda-gpu-arch=sm_70 -DACQ_REL
+// RUN: %HOST_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+
+// RUN: %clangxx -fsycl -fsycl-unnamed-lambda -fsycl-targets=%sycl_triple %s -o %t.out -Xsycl-target-backend=nvptx64-nvidia-cuda --cuda-gpu-arch=sm_70 -DACQ_REL -DATOMIC64
+// RUN: %HOST_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+
+// RUN: %clangxx -fsycl -fsycl-unnamed-lambda -fsycl-targets=%sycl_triple %s -o %t.out -Xsycl-target-backend=nvptx64-nvidia-cuda --cuda-gpu-arch=sm_70 -DACQ_REL -DWORK_GROUP
+// RUN: %HOST_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+
+// RUN: %clangxx -fsycl -fsycl-unnamed-lambda -fsycl-targets=%sycl_triple %s -o %t.out -Xsycl-target-backend=nvptx64-nvidia-cuda --cuda-gpu-arch=sm_70 -DACQ_REL -DWORK_GROUP -DATOMIC64
+// RUN: %HOST_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+
+// RUN: %clangxx -fsycl -fsycl-unnamed-lambda -fsycl-targets=%sycl_triple %s -o %t.out -Xsycl-target-backend=nvptx64-nvidia-cuda --cuda-gpu-arch=sm_70 -DACQ_REL -DSYSTEM
+// RUN: %HOST_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+
+// RUN: %clangxx -fsycl -fsycl-unnamed-lambda -fsycl-targets=%sycl_triple %s -o %t.out -Xsycl-target-backend=nvptx64-nvidia-cuda --cuda-gpu-arch=sm_70 -DACQ_REL -DSYSTEM -DATOMIC64
 // RUN: %HOST_RUN_PLACEHOLDER %t.out
 // RUN: %GPU_RUN_PLACEHOLDER %t.out
 // RUN: %CPU_RUN_PLACEHOLDER %t.out
 // RUN: %ACC_RUN_PLACEHOLDER %t.out
 
 // CUDA backend has had no support for the generic address space yet
-// XFAIL: cuda || hip
+// XFAIL: cuda || hip || level_zero || opencl
```

Comment:
Comment only mentions why CUDA is disabled. Why are the other backends disabled? (Same applies to a handful of other tests)

Reply:
It is the same reason as for CUDA. Will update the comments.

Reply:
Actually ... I might be wrong about that. I need to look into this a bit more.

Reply:
I have similar concern. Today atomic test are split into two parts: <supported everywhere> + generic pointers part not supported on HIP and CUDA backends. 
This separation allows us:
- to have full validation of all atomic features on OpenCL and Level Zero backends
- make sure that everything except atomics for generic pointers works on HIP and CUDA
- (minor) test that atomics for generic pointers fail on HIP and CUDA

New version doesn't have such clear separation. Let's look at "add" operation.

Test | Current | New
----- | ----- | -----
add.cpp | everything works everywhere | XFAIL on host, opencl and hip
add_generic.cpp | XFAIL on hip and cuda | XFAIL on hip, cuda, opencl and level_zero
add_atomic64.cpp | everything works everywhere | moved to add.cpp
add_atomic64_generic.cpp | XFAIL on hip and cuda | moved to add_generic.cpp

With new test logic distribution, we don't know if anything is working on OpenCL backends, whereas all test cases are passing. It's because new tests for floating point atomics were added to all test files and they are not supported by OpenCL. In general having multiple checks in a single test file, where at least one of the checks fails, XFAIL hides the status of all successful checks. That was the reason to move tests for "generic" pointer to a separate test.
I think it make sense to do for a new test cases which do not pass on all backends.

@t4c1, what do you think?

Reply:
I believe I might be able to resolve some of these XFAILs. Ones for floating point atomics (on add, sub, min and max) and for missing barrier on host will stay.

This PR is adding many more different subsets of tests that could be put into separate files. Previously there were no tests for atomics in local memory, atomics with non-relaxed memory orders, atomics with non-device scopes. If I followed the current convention and put every combination in a separate file it would get us a huge number of files (global/local generic AS, 32b/64b,  3 scopes, 4 orders, totaling 64 files per operation). 

We already discussed it here: https://github.com/intel/llvm-test-suite/pull/534#discussion_r743537200 and as far as I understood this was deemed an acceptable solution. If, however, you have a better suggestion how to handle this, I am happy to change it.

Reply:
I'm okay to add new test cases to existing test (e.g. add.cpp) as long as they doesn't change the test status i.e. the test should pass everywhere. 

> If I followed the current convention and put every combination in a separate file it would get us a huge number of files (global/local generic/AS, 32b/64b, 3 scopes, 4 orders, totaling 64 files per operation).

I think we can put test for features with the same validation status into the same file. Following this logic, I don't think we need a huge number of files. I expect to have 1-2 files per operation. Right?

Reply:
Hmm, we would still need to split the 3 address spaces and floating point operations, leaving us with 6 files (or 3 for operations without floating point). I guess that is still manageable.